### PR TITLE
cephfs-mirror: peer bootstrap

### PR DESCRIPTION
**Pull request by vshankar**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by vshankar** [comment]
@leseb @sebastian-philipp -- this PR introduces `peer_bootstrap {create|import}` interface to bootstrap a peer and not require peer cluster ceph configuration and keyring to be available in the primary cluster hosts.

**Comment by vshankar** [comment]
@batrick PR updated.

Test run: https://pulpito.ceph.com/vshankar-2021-03-09_09:18:41-fs-wip-cephfs-mirror-peer-bootstrap-20210309-130501-distro-basic-gibba/

File: doc/dev/cephfs-mirroring.rst [hunk 1]
```diff
@@ -32,7 +32,7 @@ Creating Users
 Start by creating a user (on the primary/local cluster) for the mirror daemon. This user
 has restrictive capabilities on the MDS and the OSD::
 
-  $ ceph auth get-or-create client.mirror mon 'allow r' mds 'allow r' osd 'allow rw tag cephfs metadata=*, allow r tag cephfs data=*' mgr 'allow r'
+  $ ceph auth get-or-create client.mirror mon 'profile cephfs-mirror' mds 'allow r' osd 'allow rw tag cephfs metadata=*, allow r tag cephfs data=*' mgr 'allow r'
```
Comment (batrick) [inline doc/dev/cephfs-mirroring.rst:35]: I know it's not new behavior with this PR but can you add a note why `cephfs-mirror` needs write permission with the metadata pool.

File: doc/dev/cephfs-mirroring.rst [hunk 2]
```diff
@@ -130,10 +130,15 @@ when added. See `Creating Users` section on how to create Ceph users for mirrori
...
 To add a peer use::
 
-  $ ceph fs snapshot mirror peer_add <fs> <remote_cluster_spec> [<remote_fs_name>]
+  $ ceph fs snapshot mirror peer_add <fs> <remote_cluster_spec> [<remote_fs_name>] [<remote_mon_host>] [<cephx_key>]
 
 `<remote_fs_name>` is optional, and default to `<fs>` (on the remote cluster).
 
+This requires the remote cluster ceph configuration and user keyring to be available in
+the primary cluster. See `Bootstrap Peers` section to avoid this. `peer_add` additionally
+suuports passing the remote cluster monitor address and the user key. However, bootstrapping
```
Comment (batrick) [inline doc/dev/cephfs-mirroring.rst:139]:
```suggestion
supports passing the remote cluster monitor address and the user key. However, bootstrapping
```

File: doc/dev/cephfs-mirroring.rst [hunk 3]
```diff
@@ -170,6 +179,29 @@ disallowed to be added for mirorring::
...
 
+Bootstrap Peers
+---------------
+
+Adding a peer (via `peer_add`) requires the peer cluster configuration and user keyring
+to be available in the primary cluster (manager host and hosts running the mirror daemon).
+This can be avoided by bootstrapping and importing a peer token. Peer bootstrap involves
+creating a bootstrap token on the peer cluster via::
+
+  $ ceph fs snapshot mirror peer_bootstrap create <fs_name> <client_entity> <site-name>
```
Comment (batrick) [inline doc/dev/cephfs-mirroring.rst:190]: Can you have one `client_entity` for multiple file systems on the the remote site?

Docs could be more clear about the meaning of `<site-name>`. What is it used for?

**Review by batrick — CHANGES_REQUESTED** [review]
otherwise LGTM; good work!

File: doc/dev/cephfs-mirroring.rst (hunk 1 above)
Comment (vshankar) [reply doc/dev/cephfs-mirroring.rst:35]: ACK

File: doc/dev/cephfs-mirroring.rst (hunk 3 above)
Comment (vshankar) [reply doc/dev/cephfs-mirroring.rst:190]: > Can you have one `client_entity` for multiple file systems on the the remote site?

The bootstrap create command creates user via `fs authorize` command and requires a file system name. One can create a user via `fs authorize \*` and use `peer_add` to reuse one user for multiple file systems.

So, `peer_bootstrap import` kind of enforces 1:1 map between a user and a filesystem.

> 
> Docs could be more clear about the meaning of `<site-name>`. What is it used for?

Sure. I'll update the docs.

**Comment by vshankar** [comment]
https://pulpito.ceph.com/vshankar-2021-03-11_07:14:32-fs-wip-cephfs-mirror-peer-bootstrap-20210310-233255-distro-basic-gibba/

**Review by batrick — APPROVED** [review]

**Pull request merged**